File: backend/app/schemas/statistics.py
# ABOUTME: Pydantic schemas for enhanced statistics API responses
# ABOUTME: Defines data structures for player statistics, head-to-head records, and performance analysis

from datetime import datetime

from pydantic import BaseModel


class HeadToHeadRecord(BaseModel):
    """Head-to-head record between two players"""

    player1_id: int
    player1_name: str
    player2_id: int
    player2_name: str
    total_games: int
    player1_wins: int
    player2_wins: int
    player1_win_percentage: float
    player2_win_percentage: float
    last_game_date: datetime | None = None
    current_streak: str | None = None  # e.g., "Player 1 - 3 game win streak"


class GameForm(BaseModel):
    """Recent form data for a single game"""

    game_id: int
    date: datetime
    opponent_id: int
    opponent_name: str
    result: str  # "W" or "L"
    rating_change: float
    conservative_rating_change: float


class RecentForm(BaseModel):
    """Recent form statistics for a player"""

    player_id: int
    player_name: str
    games_analyzed: int
    wins: int
    losses: int
    win_percentage: float
    avg_rating_change: float
    current_form: str  # e.g., "WWLWL"
    form_trend: str  # "improving", "declining", "stable"
    games: list[GameForm]


class PerformanceTrend(BaseModel):
    """Performance trend analysis over time periods"""

    period: str  # "7d", "30d", "90d", "all"
    games_played: int
    wins: int
    losses: int
    win_percentage: float
    avg_rating: float
    rating_change: float
    trend_direction: str  # "up", "down", "stable"


class PlayerStatistics(BaseModel):
    """Comprehensive player statistics"""

    player_id: int
    player_name: str

    # Basic stats
    total_games: int
    wins: int
    losses: int
    win_percentage: float

    # Rating stats
    current_mu: float
    current_sigma: float
    conservative_rating: float
    peak_rating: float
    peak_rating_date: datetime | None = None

    # Streak stats
    current_streak: str
    longest_win_streak: int
    longest_loss_streak: int

    # Performance trends
    performance_trends: list[PerformanceTrend]

    # Recent form
    recent_form: RecentForm

    # Activity stats
    first_game_date: datetime | None = None
    last_game_date: datetime | None = None
    games_this_week: int
    games_this_month: int


class PlayerStatisticsListResponse(BaseModel):
    """Response schema for paginated player statistics list"""

    statistics: list[PlayerStatistics]
    total: int
    page: int
    page_size: int
    total_pages: int


class HeadToHeadResponse(BaseModel):
    """Response schema for head-to-head statistics"""

    head_to_head: HeadToHeadRecord
    recent_games: list[GameForm]  # Last 10 games between these players


class LeaderboardEntry(BaseModel):
    """Enhanced leaderboard entry with additional statistics"""

    rank: int
    player_id: int
    player_name: str

    # Core stats
    games_played: int
    wins: int
    losses: int
    win_percentage: float

    # TrueSkill
    conservative_rating: float
    trueskill_mu: float
    trueskill_sigma: float

    # Performance indicators
    recent_form: str  # Last 5 games, e.g., "WWLWL"
    trend_7d: str  # "up", "down", "stable"
    rating_change_7d: float

    # Activity
    last_game_date: datetime | None = None
    games_this_week: int


class EnhancedLeaderboardResponse(BaseModel):
    """Enhanced leaderboard with comprehensive statistics"""

    leaderboard: list[LeaderboardEntry]
    total_players: int
    active_players: int
    total_games: int
    last_updated: datetime


class StatisticsSummary(BaseModel):
    """Overall statistics summary for the application"""

    total_players: int
    active_players: int
    total_games: int
    games_today: int
    games_this_week: int
    games_this_month: int

    # Top performers
    highest_rated_player: LeaderboardEntry | None = None
    most_active_player: LeaderboardEntry | None = None
    best_win_rate_player: LeaderboardEntry | None = None  # Min 10 games

    # System stats
    avg_games_per_player: float
    avg_rating: float
    most_common_matchup: str | None = None

    last_updated: datetime

File: frontend/src/types/statistics.ts
// ABOUTME: TypeScript type definitions for enhanced statistics features
// ABOUTME: Defines interfaces for player statistics, head-to-head records, and performance analytics

export interface HeadToHeadRecord {
  player1_id: number
  player1_name: string
  player2_id: number
  player2_name: string
  total_games: number
  player1_wins: number
  player2_wins: number
  player1_win_percentage: number
  player2_win_percentage: number
  last_game_date?: string
  current_streak?: string
}

export interface GameForm {
  game_id: number
  date: string
  opponent_id: number
  opponent_name: string
  result: 'W' | 'L'
  rating_change: number
  conservative_rating_change: number
}

export interface RecentForm {
  player_id: number
  player_name: string
  games_analyzed: number
  wins: number
  losses: number
  win_percentage: number
  avg_rating_change: number
  current_form: string // e.g., "WWLWL"
  form_trend: 'improving' | 'declining' | 'stable'
  games: GameForm[]
}

export interface PerformanceTrend {
  period: '7d' | '30d' | '90d' | 'all'
  games_played: number
  wins: number
  losses: number
  win_percentage: number
  avg_rating: number
  rating_change: number
  trend_direction: 'up' | 'down' | 'stable'
}

export interface PlayerStatistics {
  player_id: number
  player_name: string

  // Basic stats
  total_games: number
  wins: number
  losses: number
  win_percentage: number

  // Rating stats
  current_mu: number
  current_sigma: number
  conservative_rating: number
  peak_rating: number
  peak_rating_date?: string

  // Streak stats
  current_streak: string
  longest_win_streak: number
  longest_loss_streak: number

  // Performance trends
  performance_trends: PerformanceTrend[]

  // Recent form
  recent_form: RecentForm

  // Activity stats
  first_game_date?: string
  last_game_date?: string
  games_this_week: number
  games_this_month: number
}

export interface HeadToHeadResponse {
  head_to_head: HeadToHeadRecord
  recent_games: GameForm[]
}

export interface LeaderboardEntry {
  rank: number
  player_id: number
  player_name: string

  // Core stats
  games_played: number
  wins: number
  losses: number
  win_percentage: number

  // TrueSkill
  conservative_rating: number
  trueskill_mu: number
  trueskill_sigma: number

  // Performance indicators
  recent_form: string // Last 5 games, e.g., "WWLWL"
  trend_7d: 'up' | 'down' | 'stable'
  rating_change_7d: number

  // Activity
  last_game_date?: string
  games_this_week: number
}

export interface EnhancedLeaderboardResponse {
  leaderboard: LeaderboardEntry[]
  total_players: number
  active_players: number
  total_games: number
  last_updated: string
}

export interface StatisticsSummary {
  total_players: number
  active_players: number
  total_games: number
  games_today: number
  games_this_week: number
  games_this_month: number

  // Top performers
  highest_rated_player?: LeaderboardEntry
  most_active_player?: LeaderboardEntry
  best_win_rate_player?: LeaderboardEntry

  // System stats
  avg_games_per_player: number
  avg_rating: number
  most_common_matchup?: string

  last_updated: string
}

// Request parameters
export interface LeaderboardParams {
  page?: number
  page_size?: number
  min_games?: number
  sort_by?: 'rating' | 'wins' | 'games' | 'win_rate'
}

// UI-specific types
export interface StatsDashboardTab {
  id: string
  label: string
  icon?: string
}

export interface PerformanceIndicator {
  label: string
  value: string | number
  trend?: 'up' | 'down' | 'stable'
  color?: 'green' | 'red' | 'yellow' | 'blue'
  tooltip?: string
}
